### ISSUE TITLE
Unable to install beachmat (release) on linux

### ISSUE THREAD

**Issue by drisso** [tatami-inc/beachmat#7]
Hi Aaron,

I'm trying to install the Bioc release version of beachmat on a Linux server and I get the following error:

```{r}
> biocLite("beachmat")
BioC_mirror: https://bioconductor.org
Using Bioconductor 3.6 (BiocInstaller 1.28.0), R 3.4.3 (2017-11-30).
Installing package(s) ‘beachmat’
trying URL 'https://bioconductor.org/packages/3.6/bioc/src/contrib/beachmat_1.0.2.tar.gz'
Content type 'application/x-gzip' length 75131 bytes (73 KB)
==================================================
downloaded 73 KB

* installing *source* package ‘beachmat’ ...
** libs
/pbtech_mounts/softlib001/apps/EL6/spack/opt/spack/linux-rhel6-x86_64/gcc-4.4.6/gcc-6.3.0-han2ovjpjdnuwkhtfhugdq66cjkte7q3/bin/g++ -std=gnu++11 -I/pbtech_mounts/softlib001/apps/EL6/spack/opt/spack/linux-rhel6-x86_64/gcc-6.3.0/r-3.4.3-76ojsvucyrcj26lw2vjaeehyf57nlm6q/rlib/R/include -DNDEBUG  -I"/pbtech_mounts/homes029/dar2062/R/x86_64-pc-linux-gnu-library/3.4/Rcpp/include" -I"/pbtech_mounts/homes029/dar2062/R/x86_64-pc-linux-gnu-library/3.4/Rhdf5lib/include"    -fpic   -c HDF5_utils.cpp -o HDF5_utils.o
/pbtech_mounts/softlib001/apps/EL6/spack/opt/spack/linux-rhel6-x86_64/gcc-4.4.6/gcc-6.3.0-han2ovjpjdnuwkhtfhugdq66cjkte7q3/bin/g++ -std=gnu++11 -I/pbtech_mounts/softlib001/apps/EL6/spack/opt/spack/linux-rhel6-x86_64/gcc-6.3.0/r-3.4.3-76ojsvucyrcj26lw2vjaeehyf57nlm6q/rlib/R/include -DNDEBUG  -I"/pbtech_mounts/homes029/dar2062/R/x86_64-pc-linux-gnu-library/3.4/Rcpp/include" -I"/pbtech_mounts/homes029/dar2062/R/x86_64-pc-linux-gnu-library/3.4/Rhdf5lib/include"    -fpic   -c any_matrix.cpp -o any_matrix.o
/pbtech_mounts/softlib001/apps/EL6/spack/opt/spack/linux-rhel6-x86_64/gcc-4.4.6/gcc-6.3.0-han2ovjpjdnuwkhtfhugdq66cjkte7q3/bin/g++ -std=gnu++11 -I/pbtech_mounts/softlib001/apps/EL6/spack/opt/spack/linux-rhel6-x86_64/gcc-6.3.0/r-3.4.3-76ojsvucyrcj26lw2vjaeehyf57nlm6q/rlib/R/include -DNDEBUG  -I"/pbtech_mounts/homes029/dar2062/R/x86_64-pc-linux-gnu-library/3.4/Rcpp/include" -I"/pbtech_mounts/homes029/dar2062/R/x86_64-pc-linux-gnu-library/3.4/Rhdf5lib/include"    -fpic   -c character_matrix.cpp -o character_matrix.o
/pbtech_mounts/softlib001/apps/EL6/spack/opt/spack/linux-rhel6-x86_64/gcc-4.4.6/gcc-6.3.0-han2ovjpjdnuwkhtfhugdq66cjkte7q3/bin/g++ -std=gnu++11 -I/pbtech_mounts/softlib001/apps/EL6/spack/opt/spack/linux-rhel6-x86_64/gcc-6.3.0/r-3.4.3-76ojsvucyrcj26lw2vjaeehyf57nlm6q/rlib/R/include -DNDEBUG  -I"/pbtech_mounts/homes029/dar2062/R/x86_64-pc-linux-gnu-library/3.4/Rcpp/include" -I"/pbtech_mounts/homes029/dar2062/R/x86_64-pc-linux-gnu-library/3.4/Rhdf5lib/include"    -fpic   -c character_output.cpp -o character_output.o
/pbtech_mounts/softlib001/apps/EL6/spack/opt/spack/linux-rhel6-x86_64/gcc-4.4.6/gcc-6.3.0-han2ovjpjdnuwkhtfhugdq66cjkte7q3/bin/g++ -std=gnu++11 -I/pbtech_mounts/softlib001/apps/EL6/spack/opt/spack/linux-rhel6-x86_64/gcc-6.3.0/r-3.4.3-76ojsvucyrcj26lw2vjaeehyf57nlm6q/rlib/R/include -DNDEBUG  -I"/pbtech_mounts/homes029/dar2062/R/x86_64-pc-linux-gnu-library/3.4/Rcpp/include" -I"/pbtech_mounts/homes029/dar2062/R/x86_64-pc-linux-gnu-library/3.4/Rhdf5lib/include"    -fpic   -c find_chunks.cpp -o find_chunks.o
/pbtech_mounts/softlib001/apps/EL6/spack/opt/spack/linux-rhel6-x86_64/gcc-4.4.6/gcc-6.3.0-han2ovjpjdnuwkhtfhugdq66cjkte7q3/bin/g++ -std=gnu++11 -I/pbtech_mounts/softlib001/apps/EL6/spack/opt/spack/linux-rhel6-x86_64/gcc-6.3.0/r-3.4.3-76ojsvucyrcj26lw2vjaeehyf57nlm6q/rlib/R/include -DNDEBUG  -I"/pbtech_mounts/homes029/dar2062/R/x86_64-pc-linux-gnu-library/3.4/Rcpp/include" -I"/pbtech_mounts/homes029/dar2062/R/x86_64-pc-linux-gnu-library/3.4/Rhdf5lib/include"    -fpic   -c functions.cpp -o functions.o
/pbtech_mounts/softlib001/apps/EL6/spack/opt/spack/linux-rhel6-x86_64/gcc-4.4.6/gcc-6.3.0-han2ovjpjdnuwkhtfhugdq66cjkte7q3/bin/g++ -std=gnu++11 -I/pbtech_mounts/softlib001/apps/EL6/spack/opt/spack/linux-rhel6-x86_64/gcc-6.3.0/r-3.4.3-76ojsvucyrcj26lw2vjaeehyf57nlm6q/rlib/R/include -DNDEBUG  -I"/pbtech_mounts/homes029/dar2062/R/x86_64-pc-linux-gnu-library/3.4/Rcpp/include" -I"/pbtech_mounts/homes029/dar2062/R/x86_64-pc-linux-gnu-library/3.4/Rhdf5lib/include"    -fpic   -c integer_matrix.cpp -o integer_matrix.o
/pbtech_mounts/softlib001/apps/EL6/spack/opt/spack/linux-rhel6-x86_64/gcc-4.4.6/gcc-6.3.0-han2ovjpjdnuwkhtfhugdq66cjkte7q3/bin/g++ -std=gnu++11 -I/pbtech_mounts/softlib001/apps/EL6/spack/opt/spack/linux-rhel6-x86_64/gcc-6.3.0/r-3.4.3-76ojsvucyrcj26lw2vjaeehyf57nlm6q/rlib/R/include -DNDEBUG  -I"/pbtech_mounts/homes029/dar2062/R/x86_64-pc-linux-gnu-library/3.4/Rcpp/include" -I"/pbtech_mounts/homes029/dar2062/R/x86_64-pc-linux-gnu-library/3.4/Rhdf5lib/include"    -fpic   -c logical_matrix.cpp -o logical_matrix.o
/pbtech_mounts/softlib001/apps/EL6/spack/opt/spack/linux-rhel6-x86_64/gcc-4.4.6/gcc-6.3.0-han2ovjpjdnuwkhtfhugdq66cjkte7q3/bin/g++ -std=gnu++11 -I/pbtech_mounts/softlib001/apps/EL6/spack/opt/spack/linux-rhel6-x86_64/gcc-6.3.0/r-3.4.3-76ojsvucyrcj26lw2vjaeehyf57nlm6q/rlib/R/include -DNDEBUG  -I"/pbtech_mounts/homes029/dar2062/R/x86_64-pc-linux-gnu-library/3.4/Rcpp/include" -I"/pbtech_mounts/homes029/dar2062/R/x86_64-pc-linux-gnu-library/3.4/Rhdf5lib/include"    -fpic   -c numeric_matrix.cpp -o numeric_matrix.o
/pbtech_mounts/softlib001/apps/EL6/spack/opt/spack/linux-rhel6-x86_64/gcc-4.4.6/gcc-6.3.0-han2ovjpjdnuwkhtfhugdq66cjkte7q3/bin/g++ -std=gnu++11 -I/pbtech_mounts/softlib001/apps/EL6/spack/opt/spack/linux-rhel6-x86_64/gcc-6.3.0/r-3.4.3-76ojsvucyrcj26lw2vjaeehyf57nlm6q/rlib/R/include -DNDEBUG  -I"/pbtech_mounts/homes029/dar2062/R/x86_64-pc-linux-gnu-library/3.4/Rcpp/include" -I"/pbtech_mounts/homes029/dar2062/R/x86_64-pc-linux-gnu-library/3.4/Rhdf5lib/include"    -fpic   -c output_param.cpp -o output_param.o
/pbtech_mounts/softlib001/apps/EL6/spack/opt/spack/linux-rhel6-x86_64/gcc-4.4.6/gcc-6.3.0-han2ovjpjdnuwkhtfhugdq66cjkte7q3/bin/g++ -std=gnu++11 -I/pbtech_mounts/softlib001/apps/EL6/spack/opt/spack/linux-rhel6-x86_64/gcc-6.3.0/r-3.4.3-76ojsvucyrcj26lw2vjaeehyf57nlm6q/rlib/R/include -DNDEBUG  -I"/pbtech_mounts/homes029/dar2062/R/x86_64-pc-linux-gnu-library/3.4/Rcpp/include" -I"/pbtech_mounts/homes029/dar2062/R/x86_64-pc-linux-gnu-library/3.4/Rhdf5lib/include"    -fpic   -c repacker.cpp -o repacker.o
/pbtech_mounts/softlib001/apps/EL6/spack/opt/spack/linux-rhel6-x86_64/gcc-4.4.6/gcc-6.3.0-han2ovjpjdnuwkhtfhugdq66cjkte7q3/bin/g++ -std=gnu++11 -I/pbtech_mounts/softlib001/apps/EL6/spack/opt/spack/linux-rhel6-x86_64/gcc-6.3.0/r-3.4.3-76ojsvucyrcj26lw2vjaeehyf57nlm6q/rlib/R/include -DNDEBUG  -I"/pbtech_mounts/homes029/dar2062/R/x86_64-pc-linux-gnu-library/3.4/Rcpp/include" -I"/pbtech_mounts/homes029/dar2062/R/x86_64-pc-linux-gnu-library/3.4/Rhdf5lib/include"    -fpic   -c utils.cpp -o utils.o
/pbtech_mounts/softlib001/apps/EL6/spack/opt/spack/linux-rhel6-x86_64/gcc-4.4.6/gcc-6.3.0-han2ovjpjdnuwkhtfhugdq66cjkte7q3/bin/g++ -std=gnu++11 -shared -L/pbtech_mounts/softlib001/apps/EL6/spack/opt/spack/linux-rhel6-x86_64/gcc-6.3.0/r-3.4.3-76ojsvucyrcj26lw2vjaeehyf57nlm6q/rlib/R/lib -o beachmat.so HDF5_utils.o any_matrix.o character_matrix.o character_output.o find_chunks.o functions.o integer_matrix.o logical_matrix.o numeric_matrix.o output_param.o repacker.o utils.o /pbtech_mounts/homes029/dar2062/R/x86_64-pc-linux-gnu-library/3.4/Rhdf5lib/lib/libhdf5_cpp.a /pbtech_mounts/homes029/dar2062/R/x86_64-pc-linux-gnu-library/3.4/Rhdf5lib/lib/libhdf5.a /pbtech_mounts/homes029/dar2062/R/x86_64-pc-linux-gnu-library/3.4/Rhdf5lib/lib/libsz.a -L/pbtech_mounts/softlib001/apps/EL6/spack/opt/spack/linux-rhel6-x86_64/gcc-6.3.0/r-3.4.3-76ojsvucyrcj26lw2vjaeehyf57nlm6q/rlib/R/lib -lR
ar rc libbeachmat.a any_matrix.o character_matrix.o character_output.o integer_matrix.o logical_matrix.o numeric_matrix.o utils.o HDF5_utils.o output_param.o
/pbtech_mounts/softlib001/apps/EL6/spack/opt/spack/linux-rhel6-x86_64/gcc-4.4.6/gcc-6.3.0-han2ovjpjdnuwkhtfhugdq66cjkte7q3/bin/g++ -std=gnu++11 -shared  -o libbeachmat.so any_matrix.o character_matrix.o character_output.o integer_matrix.o logical_matrix.o numeric_matrix.o utils.o HDF5_utils.o output_param.o `echo 'Rhdf5lib::pkgconfig("PKG_CXX_LIBS")'| "/pbtech_mounts/softlib001/apps/EL6/spack/opt/spack/linux-rhel6-x86_64/gcc-6.3.0/r-3.4.3-76ojsvucyrcj26lw2vjaeehyf57nlm6q/rlib/R/bin/R" --vanilla --slave` 
mkdir -p "/pbtech_mounts/homes029/dar2062/R/x86_64-pc-linux-gnu-library/3.4/beachmat/lib" "/pbtech_mounts/homes029/dar2062/R/x86_64-pc-linux-gnu-library/3.4/beachmat/include/beachmat"
cp any_matrix.h utils.h beachmat.h HDF5_utils.h output_param.h Psymm_matrix.h HDF5_matrix.h Csparse_matrix.h dense_matrix.h simple_matrix.h Rle_matrix.h Input_matrix.h simple_output.h Csparse_output.h HDF5_output.h Output_matrix.h LIN_matrix.h LIN_methods.h LIN_output.h LIN_outfun.h logical_matrix.h integer_matrix.h character_matrix.h numeric_matrix.h character_output.h   "/pbtech_mounts/homes029/dar2062/R/x86_64-pc-linux-gnu-library/3.4/beachmat/include/beachmat"
mv libbeachmat.* "/pbtech_mounts/homes029/dar2062/R/x86_64-pc-linux-gnu-library/3.4/beachmat/lib"
installing to /pbtech_mounts/homes029/dar2062/R/x86_64-pc-linux-gnu-library/3.4/beachmat/libs
** R
** inst
** preparing package for lazy loading
** help
*** installing help indices
** building package indices
** installing vignettes
** testing if installed package can be loaded
Error: package or namespace load failed for ‘beachmat’ in dyn.load(file, DLLpath = DLLpath, ...):
 unable to load shared object '/pbtech_mounts/homes029/dar2062/R/x86_64-pc-linux-gnu-library/3.4/beachmat/libs/beachmat.so':
  /pbtech_mounts/homes029/dar2062/R/x86_64-pc-linux-gnu-library/3.4/beachmat/libs/beachmat.so: undefined symbol: _ZN2H56H5FileC1ERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEjRKNS_17FileCreatPropListERKNS_15FileAccPropListE
Error: loading failed
Execution halted
ERROR: loading failed
* removing ‘/pbtech_mounts/homes029/dar2062/R/x86_64-pc-linux-gnu-library/3.4/beachmat’
```

Any idea what's going wrong?

This is my `sessionInfo()` in case you find it useful:

```{r}
R version 3.4.3 (2017-11-30)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: CentOS release 6.8 (Final)

Matrix products: default
BLAS: /pbtech_mounts/softlib001/apps/EL6/spack/opt/spack/linux-rhel6-x86_64/gcc-6.3.0/r-3.4.3-76ojsvucyrcj26lw2vjaeehyf57nlm6q/rlib/R/lib/libRblas.so
LAPACK: /pbtech_mounts/softlib001/apps/EL6/spack/opt/spack/linux-rhel6-x86_64/gcc-6.3.0/r-3.4.3-76ojsvucyrcj26lw2vjaeehyf57nlm6q/rlib/R/lib/libRlapack.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] Rhdf5lib_1.0.0       BiocInstaller_1.28.0

loaded via a namespace (and not attached):
[1] compiler_3.4.3 tools_3.4.3   

```

**Comment by LTLA** [comment]
It's the old nemesis - see the "Common problems" section of the new BioC-devel vignette at:

https://bioconductor.org/packages/devel/bioc/vignettes/beachmat/inst/doc/linking.html

I'm guessing you're using some non-system package manager, which probably upgraded your GCC in between _beachmat_ and _Rhdf5lib_ installations while you weren't looking.

**Comment by drisso** [comment]
Ah... it's even more evil than that... for some reason my system is using a different version of GCC for the two packages even when I'm installing them in the same R session... thanks for pointing me in the right direction!

From Rhdf5lib:
```
Languages:
----------
                        Fortran: no

                            C++: yes
                   C++ Compiler: /usr/bin/g++ ( g++ (GCC) 4.4.7 20120313 )
                      C++ Flags: 
                   H5 C++ Flags:  
                   AM C++ Flags: 
             Shared C++ Library: no
             Static C++ Library: yes

```

While beachmat is using:
```
/pbtech_mounts/softlib001/apps/EL6/spack/opt/spack/linux-rhel6-x86_64/gcc-4.4.6/gcc-6.3.0-han2ovjpjdnuwkhtfhugdq66cjkte7q3/bin/g++
```

**Comment by drisso** [comment]
I'm adding this just in the remote case that someone with exactly my set up runs into this problem (or more likely when I hit the same error in the future!).

My system uses Spack as a package manager. I was not explicitly loading any version of GCC when I hit the error above. The problem is solved if I first load gcc-6.3.0 before installing the packages.
